### PR TITLE
Adjust feature flag logic for cody on dotcom

### PR DESCRIPTION
**Pull request by eseliger**
This PR changes to no access to unauthenticated users. We want to hide cody on dotcom. I've also added tests to verify things behave alright. The feature flag was currently not correct. The backend made a different determination than the frontend did, because the frontend simply looked a the feature flag, while the backend looks at the `ExperimentalFeatures.CodyRestrictUsersFeatureFlag` field as well. This aligns the two for now by exposing codyEnabled on the jscontext and using the same logic in the backend for that. This is likely not the end state we want here, but it fixes an immediate problem where unauthed users would be able to see Cody but then aren't able to use it.

## Test plan

Added tests, verified manually.

### CODE REVIEW

**Comment by sourcegraph-buildkite** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.00% (-0.03 kb) | -0.00% (-0.07 kb) | -0.00% (-0.04 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/es/ff-cody/compare-report.html;tab=live_object) for a full comparison between the commits bf41af8 and ff90add or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: internal/cody/feature_flag.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/featureflag"
+)
+
+// IsCodyEnabled determines if cody is enabled for the actor in the given context.
+// If it is an unauthenticated request, cody is disabled.
+// Otherwise, if CodyRestrictUsersFeatureFlag is set, the cody-experimental featureflag
+// will determine access.
+// Elseway, all authenticated users are granted access.
```
Comment (vdavid) [inline internal/cody/feature_flag.go:15]: Nit: I don't think "elseway" is a word? :D

File: internal/cody/feature_flag_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"context"
+	"testing"
+
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/featureflag"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+func TestIsCodyEnabled(t *testing.T) {
```
Comment (vdavid) [inline internal/cody/feature_flag_test.go:13]: Nit: I'd break this up to multiple functions for clarity. I'd love to have names for each test case.

**Review by vdavid — APPROVED** [review]
Left two nits. I love this simplification and untangling, plus that it's better tested now.

File: internal/cody/feature_flag.go (hunk 1 above)
Comment (eseliger) [reply internal/cody/feature_flag.go:15]: lol, according to google it's dialectal

File: internal/cody/feature_flag_test.go (hunk 2 above)
Comment (eseliger) [reply internal/cody/feature_flag_test.go:13]: will do 👍

**Review by malomarrec — APPROVED** [review]

**Comment by malomarrec** [comment]
Approved, but recommend considering Taylor as a blocking reviewer before merge to be on the safe side

**Comment by eseliger** [comment]
> Because any visitor (authd or unauthd) can see Cody in the UI right now, we need to let them know they just need to sign in. 

Yep, after this PR it will be hidden for unauthed and visible for signed in

> I have a proposal for how to handle this (at least temporarily) in https://github.com/sourcegraph/sourcegraph/pull/51229. It's a simple conditional (though the condition may have changed) in [client/web/src/storm/pages/SearchPage/SearchPageContent.tsx](https://github.com/sourcegraph/sourcegraph/pull/51229/files#diff-8275bb424052602a299c76aeab2bcdb1c6867c4b7a1f088388b2193fa8b10b67). Maybe we can fold that change into this PR?

I already have the changes for that made just need to push to your branch after this is merged and we should be good. On dotcom, the regular cody UI will still be hidden then, but your notice will appear on the cody search page, like your PR suggested. 

Is that the final state we wish to have?

**Review by taylorsperry — APPROVED** [review]

**Pull request merged**